### PR TITLE
Fix performance regression of writes in JDBC connectors

### PR DESCRIPTION
**Pull request by hashhar**
This reverts commit 541edbe2225ee91ed678be1aad000248c3a0e57a.

### CODE REVIEW

File: plugin/trino-base-jdbc/src/main/java/io/trino/plugin/jdbc/JdbcPageSink.java [hunk 1]
```diff
@@ -161,6 +174,7 @@ else if (javaType == Slice.class) {
                 PreparedStatement statement = this.statement) {
             if (batchSize > 0) {
                 statement.executeBatch();
+                connection.commit();
```
Comment (hashhar) [inline plugin/trino-base-jdbc/src/main/java/io/trino/plugin/jdbc/JdbcPageSink.java:177]: @findepi Do we assume that any connection pool or the driver will ensure that the connection's `autoCommit` state it reset before giving the connection to a new client?
Should I explicitly save (in constructor) and restore (here) the autocommit state?

I think this will cause failures with Oracle UCP (it doesn't reset things AFAIK) - let's see.

Comment (hashhar) [reply plugin/trino-base-jdbc/src/main/java/io/trino/plugin/jdbc/JdbcPageSink.java:177]: I ask because I remember that for DELETEs Grzegorz is using autocommit (which makes sense because it's always a single statement there) and if the pool gives a stale connection DELETEs will fail.

Comment (hashhar) [reply plugin/trino-base-jdbc/src/main/java/io/trino/plugin/jdbc/JdbcPageSink.java:177]: Yep, it failed as expected.

And it absolutely has to do with pool returning non-reset connections because if I remove anything that uses the PageSink by applying the below diff then the test passes.

I'm fixing this by changing the OraclePoolConnectionFactory to always return connections with autoCommit enabled.

<details><summary>Click here for diff</summary>
```diff
diff --git a/plugin/trino-oracle/src/test/java/io/trino/plugin/oracle/TestOraclePoolConnectorSmokeTest.java b/plugin/trino-oracle/src/test/java/io/trino/plugin/oracle/TestOraclePoolConnectorSmokeTest.java
index 30d6108283..e8a38c8c75 100644
--- a/plugin/trino-oracle/src/test/java/io/trino/plugin/oracle/TestOraclePoolConnectorSmokeTest.java
+++ b/plugin/trino-oracle/src/test/java/io/trino/plugin/oracle/TestOraclePoolConnectorSmokeTest.java
@@ -13,15 +13,20 @@
  */
 package io.trino.plugin.oracle;
 
+import com.google.common.collect.ImmutableList;
 import com.google.common.collect.ImmutableMap;
 import io.airlift.testing.Closeables;
 import io.trino.testing.QueryRunner;
+import io.trino.testing.sql.TestTable;
 import org.testng.annotations.AfterClass;
+import org.testng.annotations.Test;
 
 import java.io.IOException;
 
 import static io.trino.plugin.oracle.TestingOracleServer.TEST_PASS;
 import static io.trino.plugin.oracle.TestingOracleServer.TEST_USER;
+import static io.trino.testing.TestingConnectorBehavior.SUPPORTS_CREATE_TABLE;
+import static io.trino.testing.TestingConnectorBehavior.SUPPORTS_DELETE;
 
 public class TestOraclePoolConnectorSmokeTest
         extends BaseOracleConnectorSmokeTest
@@ -44,7 +49,7 @@ public class TestOraclePoolConnectorSmokeTest
                         .put("oracle.connection-pool.enabled", "true")
                         .put("oracle.remarks-reporting.enabled", "false")
                         .build(),
-                REQUIRED_TPCH_TABLES);
+                ImmutableList.of());
     }
 
     @AfterClass(alwaysRun = true)
@@ -54,4 +59,16 @@ public class TestOraclePoolConnectorSmokeTest
         Closeables.closeAll(oracleServer);
         oracleServer = null;
     }
+
+    @Test
+    public void testDeleteAllDataFromTable()
+    {
+        skipTestUnless(hasBehavior(SUPPORTS_CREATE_TABLE) && hasBehavior(SUPPORTS_DELETE));
+        try (TestTable table = new TestTable(oracleServer::execute, "test_delete", "AS SELECT 'good' foo FROM DUAL")) {
+            // not using assertUpdate as some connectors provide update count and some do not
+            assertQuery("SELECT count(*) FROM " + table.getName(), "VALUES 1");
+            getQueryRunner().execute("DELETE FROM " + table.getName());
+            assertQuery("SELECT count(*) FROM " + table.getName(), "VALUES 0");
+        }
+    }
 }
```
</summary>

**Comment by kokosing** [comment]
Can you please share more about `performance regression` you are fixing here?

File: plugin/trino-base-jdbc/src/main/java/io/trino/plugin/jdbc/JdbcPageSink.java (hunk 1 above)
Comment (kokosing) [reply plugin/trino-base-jdbc/src/main/java/io/trino/plugin/jdbc/JdbcPageSink.java:177]: When was playing with Oracle, I noticed that changing a state of the connection survives in a connection in the pool.

**Comment by hashhar** [comment]
> Can you please share more about `performance regression` you are fixing here?

Sure. In https://github.com/trinodb/trino/commit/541edbe2225ee91ed678be1aad000248c3a0e57a we changed the PageSink to use autocommit instead of explicit transaction management.
They are equivalent ONLY when a single statement is being executed.
But in JdbcPageSink that's not the case. We add multiple statements (1 per `addBatch` call) and then `executeBatch`.

From JDBC javadocs "If a connection is in auto-commit mode, then all its SQL statements will be executed and committed as individual transactions". MySQL and SQL Server follow this and hence those 1000 `addBatch` calls turn into 1000 commits which kills performance.

Using explicit transaction ensures that we only do 1 COMMIT for the entire batch instead of commit per statement.

For comparision I added a test that does a CTAS and INSERT of `tpch.tiny.orders`.

On MySQL:

```
# Current master
INSERT [27598, 35814, 48281, 37213, 42254]; Avg(ms): 38232.0
CTAS [25855, 35361, 40970, 63781, 57870]; Avg(ms): 44767.4

# After revert
INSERT [16066, 14602, 12628, 11886, 14542]; Avg(ms): 13944.8
CTAS [14474, 23864, 24198, 12889, 13070]; Avg(ms): 17699.0
```

On SQL Server:

```
# Current master
INSERT [27110, 24551, 29237, 40681, 41884]; Avg(ms): 32692.6
CTAS [46106, 37056, 34141, 29358, 45614]; Avg(ms): 38455.0

# After revert
INSERT [10097, 2962, 7849, 4154, 1923]; Avg(ms): 5397.0
CTAS [10392, 7737, 7869, 2042, 3284]; Avg(ms): 6264.8
```

File: plugin/trino-base-jdbc/src/main/java/io/trino/plugin/jdbc/JdbcPageSink.java (hunk 1 above)
Comment (findepi) [reply plugin/trino-base-jdbc/src/main/java/io/trino/plugin/jdbc/JdbcPageSink.java:177]: > Do we assume that any connection pool or the driver will ensure that the connection's `autoCommit` state it reset before giving the connection to a new client?

That should be guaranteed by the pool.

Anything else is very error-prone.

> When was playing with Oracle, I noticed that changing a state of the connection survives in a connection in the pool.

Was this about autocommit state, or something else?

File: plugin/trino-base-jdbc/src/main/java/io/trino/plugin/jdbc/JdbcPageSink.java [hunk 2]
```diff
@@ -60,6 +60,11 @@ public JdbcPageSink(ConnectorSession session, JdbcOutputTableHandle handle, Jdbc
         }
 
         try {
+            // According to JDBC javaodcs "If a connection is in auto-commit mode, then all its SQL statements will be
+            // executed and committed as individual transactions." Notably MySQL and SQL Server respect this which
+            // leads to multiple commits when we close the connection leading to slow performance. Explicit commits
+            // where needed ensure that all of the submitted statements are committed as a single transaction and
+            // performs better.
```
Comment (findepi) [inline plugin/trino-base-jdbc/src/main/java/io/trino/plugin/jdbc/JdbcPageSink.java:67]: Move to separate commit, so that revert is a pure revert.

However, please add this to revert's commit message -- revert commit message should tell the "why revert".

File: plugin/trino-oracle/src/main/java/io/trino/plugin/oracle/OraclePoolConnectionFactory.java [hunk 3]
```diff
@@ -80,6 +80,10 @@ public OraclePoolConnectionFactory(
     public Connection openConnection(ConnectorSession session)
             throws SQLException
     {
-        return dataSource.getConnection();
+        Connection connection = dataSource.getConnection();
+        // Oracle's pool doesn't reset autocommit of connections when reusing them so we explicitly enable autocommit
+        // by default to match the JDBC spec.
+        connection.setAutoCommit(true);
+        return connection;
```
Comment (findepi) [inline plugin/trino-oracle/src/main/java/io/trino/plugin/oracle/OraclePoolConnectionFactory.java:87]: That's not part of the revert commit. Make it a preparatory commit.

File: plugin/trino-oracle/src/main/java/io/trino/plugin/oracle/OraclePoolConnectionFactory.java [hunk 4]
```diff
@@ -80,6 +80,10 @@ public OraclePoolConnectionFactory(
     public Connection openConnection(ConnectorSession session)
             throws SQLException
     {
-        return dataSource.getConnection();
+        Connection connection = dataSource.getConnection();
+        // Oracle's pool doesn't reset autocommit of connections when reusing them so we explicitly enable autocommit
+        // by default to match the JDBC spec.
+        connection.setAutoCommit(true);
```
Comment (findepi) [inline plugin/trino-oracle/src/main/java/io/trino/plugin/oracle/OraclePoolConnectionFactory.java:86]: Can the oracle pool be configured to do that?

**Review by findepi — APPROVED** [review]

File: plugin/trino-oracle/src/main/java/io/trino/plugin/oracle/OraclePoolConnectionFactory.java (hunk 4 above)
Comment (hashhar) [reply plugin/trino-oracle/src/main/java/io/trino/plugin/oracle/OraclePoolConnectionFactory.java:86]: Yes, by passing the connection property `OracleConnection.CONNECTION_PROPERTY_AUTOCOMMIT` to `true` to the PoolDataSource.
I'll do that.

File: plugin/trino-base-jdbc/src/main/java/io/trino/plugin/jdbc/JdbcPageSink.java (hunk 1 above)
Comment (hashhar) [reply plugin/trino-base-jdbc/src/main/java/io/trino/plugin/jdbc/JdbcPageSink.java:177]: I observed it for autocommit state. Hence the 2nd fixup commit.

Hikari does guarantee it - Oracle UCP seems not to.

File: plugin/trino-oracle/src/main/java/io/trino/plugin/oracle/OraclePoolConnectionFactory.java (hunk 4 above)
Comment (hashhar) [reply plugin/trino-oracle/src/main/java/io/trino/plugin/oracle/OraclePoolConnectionFactory.java:86]: Not quite - turns out that it only affects the initial connection. UCP isn't touching the connections once created.

**Comment by hashhar** [comment]
@findepi @kokosing PTAL again. applied comments.

**Review by findepi — APPROVED** [review]

**Review by kokosing — APPROVED** [review]

**Comment by hashhar** [comment]
All green.
